service/notification/internal/app/scheduler/scheduler.go

```diff
@@ -0,0 +1,169 @@
+		scheduledDispatches: make(map[string]*ScheduledDispatch),
+	}
+}
+
+func (s *dispatchScheduler) invokeSendingOfDispatch(d *entities.Dispatch) {
+	msg := messages.SendDispatchCommand{
+		EventType: messages.EventType_SEND_DISPATCH,
+		Timestamp: timestamppb.New(time.Now().UTC()),
+		Payload: &messages.Dispatch{
+			Emails:  d.Emails,
+			BaseCcy: d.BaseCcy,
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), app.TimeoutOfProcessing)
+	defer cancel()
+
+	rates, err := s.converter.Convert(ctx, d.BaseCcy, d.TargetCcies)
+	if err != nil {
+		s.logger.Errorf("failed to get rates: %v", err)
+
+		return
+	}
+	msg.Payload.Rates = rates
+
+	marshalled, err := proto.Marshal(&msg)
+	if err != nil {
+		s.logger.Errorf("failed to marshal SendDispatchCommand: %v", err)
+
+		return
+	}
+
+	if err := s.broker.PublishAsync(app.SendDispatchCommand, marshalled); err != nil {
+		s.logger.Errorf("failed to emit SendDispatch commands: %v", err)
+	}
+}
+
+func (s *dispatchScheduler) Run() {
+	s.cron.Run()
+}
+
+func (s *dispatchScheduler) Stop() {
+	s.cron.Stop()
+}
+
+func (s *dispatchScheduler) AddDispatches(ds map[string]entities.Dispatch) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, dispatch := range ds {
+		scheduled := ScheduledDispatch{
+			Data: &dispatch,
+			Spec: NewJobSpec(dispatch.SendAt),
+		}
+
+		entryID, err := s.cron.AddJob(scheduled.Spec.String(), NewSendDispatchJob(scheduled.Data, s.invokeSendingOfDispatch))
+		if err != nil {
+			return
+		}
+
+		scheduled.EntryID = entryID
+
+		s.scheduledDispatches[dispatch.ID] = &scheduled
+	}
+}
+
+func (s *dispatchScheduler) AddSubscription(sub entities.Subscription) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	dispatch, ok := s.scheduledDispatches[sub.DispatchID]
+	if !ok {
+		dispatch = &ScheduledDispatch{
+			Data: sub.ToDispatch(),
+			Spec: NewJobSpec(sub.SendAt),
+		}
+
+		s.scheduledDispatches[sub.DispatchID] = dispatch
+	} else {
+		dispatch.Data.Emails = append(dispatch.Data.Emails, sub.Email)
+	}
+
+	entryID, err := s.cron.AddJob(dispatch.Spec.String(), NewSendDispatchJob(dispatch.Data, s.invokeSendingOfDispatch))
+	if err != nil {
+		return
+	}
+
+	dispatch.EntryID = entryID
+	s.scheduledDispatches[sub.DispatchID] = dispatch
+}
+
+// func (s *dispatchScheduler) CancelSubscription(sub entities.Subscription) {
+// 	s.mu.Lock()
+// 	defer s.mu.Unlock()
+// 	schdledDispatch, ok := s.scheduledDispatches[sub.DispatchID]
+// 	if !ok {
+// 		return
+// 	}
+// 	emailIndex := slices.Index(schdledDispatch.Data.Emails, sub.Email)
+// 	if emailIndex < 0 {
+// 		return
+// 	}
+// 	countOfSubscribers := len(schdledDispatch.Data.Emails)
+// 	schdledDispatch.Data.Emails[emailIndex] = schdledDispatch.Data.Emails[countOfSubscribers-1]
+// 	schdledDispatch.Data.Emails = schdledDispatch.Data.Emails[:countOfSubscribers-1]
+// 	s.cron.Remove(schdledDispatch.entryID)
+// 	s.cron.AddJob(schdledDispatch.spec.String(), NewDispatchJob(schdledDispatch.Data, s.broker))
+// }
```

Comment:
Redundant?

Reply:
In this PR it is, but in next it will be used. Just ignore it.